File: leetcode/cpp/P1282_GroupPeople.cpp
// https://leetcode.com/problems/group-the-people-given-the-group-size-they-belong-to/

#include <unordered_map>
#include <vector>

class Solution {
 public:
  std::vector<std::vector<int>> groupThePeople(std::vector<int>& groupSizes) {
    std::unordered_map<int, std::vector<int>> crt;
    std::vector<std::vector<int>> solution;

    for (int id = 0; id < groupSizes.size(); ++id) {
      int crtSize = groupSizes[id];

      if (crt[crtSize].size() < crtSize) {
        crt[crtSize].push_back(id);
      }

      if (crt[crtSize].size() == crtSize) {
        solution.push_back(crt[crtSize]);
        crt[crtSize] = std::vector<int>();
      }
    }
    return solution;
  }
};

File: leetcode/python/P1282_GroupPeople.py
# https://leetcode.com/problems/group-the-people-given-the-group-size-they-belong-to/

from typing import List, Dict


class Solution:
    def groupThePeople(self, group_sizes: List[int]) -> List[List[int]]:
        # n is small (<= 500)
        # we can return any solution, so we can fill groups
        # as they occur.
        # keep a dict size: list of elements
        # when we reach the capacity for the list, append it to the
        # solution
        result: List[List[int]] = []
        buffers: Dict[int, List] = {}
        for num, group_sz in enumerate(group_sizes):
            if buffers.get(group_sz, None) is None:
                buffers[group_sz] = []
            buffers[group_sz].append(num)
            if len(buffers[group_sz]) == group_sz:
                result.append(buffers[group_sz])
                buffers[group_sz] = []

        return result
